src/prototype.rs

```diff
@@ -72,20 +80,83 @@ pub trait Prototypical: 'static + Send + Sync {
 	) -> EntityCommands<'a, 'b> {
 		let mut entity = commands.spawn();
 		let mut proto_commands = self.create_commands(entity, data);
-		for component in self.iter_components() {
-			component.insert_self(&mut proto_commands, asset_server);
-		}
+
+		spawn_internal(
+			self.name(),
+			self.template(),
+			self.iter_components(),
+			&mut proto_commands,
+			data,
+			asset_server,
+			&mut IndexSet::default(),
+		);
 
 		proto_commands.into()
 	}
 }
 
+/// Internal method used for recursing up the template hierarchy and spawning components
+/// from the top to the bottom
+fn spawn_internal<'a>(
+	name: &'a str,
+	template: Option<&'a str>,
+	components: Iter<'a, Box<dyn ProtoComponent>>,
+	proto_commands: &mut ProtoCommands,
+	data: &'a Res<ProtoData>,
+	asset_server: &Res<AssetServer>,
+	traversed: &mut IndexSet<&'a str>,
+) {
+	// We insert first on the off chance that someone made a prototype its own template...
+	traversed.insert(name);
+
+	match template {
+		Some(template_name) if traversed.contains(template_name) => {
+			// ! === Found Circular Dependency === ! //
+			let tree: String = traversed
+				.iter()
+				.map(|n| format!("'{}' -> ", n))
+				.collect::<String>()
+				.add(&format!("'{}'", template_name));
+			warn!(
```

Comment:
Perhaps a feature flag in the Cargo config could enable whether this is considered a fatal error or not. In dev, maybe not such a big deal. In release, we might want to panic.

Reply:
That's a good idea. I've actually never done features before so this will be as good a time to learn as ever 😃

Reply:
Although, we probably won't panic here. If that feature is enabled, we'll probably panic much earlier, when we first load all the prototypes in.

Reply:
In response to this suggestion, I added two features:
https://github.com/MrGVSV/bevy_proto/blob/6f0bb2e787792d145f6da8ca8a42473d44a3663b/Cargo.toml#L22-L25

`analysis` enables the prototype analysis on load. While `no_cycles` toggles between warning (disabled) and panicking (enabled). Not sure `analysis` is super-needed, but left it in, just in case a user has a desire to disable it.